# PanScan
PanScan is a computational toolkit for the exploration of **gene duplications, novel sequences and complex structural variants (SVs)** .  
It supports **gene duplication annotation**, **complex region visualization**, and provides modules for **variant and sequence comparison**.

## ðŸš€ Quick Start (Containerized Installation)
The easiest way to use **PanScan** is with its **Docker** or **Singularity** image, both preloaded with all dependencies.

### **Option 1: Docker**
Pull the latest version:
```bash
docker pull ghcr.io/catg-github/panscan:latest
```

Run:
```bash
docker run --rm -it ghcr.io/catg-github/panscan:latest panscan
```

### **Option 2: Singularity**
Pull the image:
```bash
singularity pull ghcr.io/catg-github/panscan:latest
```

Run:
```bash
singularity exec panscan.sif panscan
```

You should see:
```
usage: panscan [-h] {complex,preprocess_vcf,novel_seq,find_uniq_variants,make_dup_mtx,gene_dup} ...
```

> ðŸ’¡ *Using Docker or Singularity removes the need to install Perl, Python, R, or other dependencies manually.*

## ðŸ§° Installation from Source
If you prefer to install from source:

```bash
git clone https://github.com/CATG-Github/panscan.git
cd panscan
pip install .
```

Run with:
```bash
panscan
```

## ðŸ“¦ Requirements (for manual installations)
**Programming Languages**
- [Perl](https://www.perl.org/)
- [Python â‰¥ 3.7](https://www.python.org/)
- [R](https://www.r-project.org/)
- [Java](https://www.java.com/en/)

**Python Packages**
`argparse, Image, ImageDraw, ImageFont, matplotlib, numpy, os, pandas, pickle, sys`

**Perl Modules**
`Getopt::Long,  Cwd, File::Copy, Exporter`

**Bioinformatics Tools**
- [Bandage](https://rrwick.github.io/Bandage)
- [BCFtools](https://github.com/samtools/bcftools)
- [bgzip](https://www.htslib.org/doc/bgzip.html)
- [cd-hit](https://github.com/weizhongli/cdhit)
- [GFABase](https://github.com/mlin/gfabase)
- [GraphAligner](https://github.com/maickrau/GraphAligner)
- [Liftoff](https://github.com/agshumate/Liftoff)
- [rtg-tools](https://github.com/RealTimeGenomics/rtg-tools)
- [tabix](https://www.htslib.org/doc/tabix.html)
- [truvari](https://github.com/ACEnglish/truvari)
- [RIdeogram](https://cran.r-project.org/web/packages/RIdeogram/vignettes/RIdeogram.html)

## **Required Databases**
[Panscan Zenodo Databases](https://zenodo.org/records/17358559)  
BED files derived from [dbSNP](https://www.ncbi.nlm.nih.gov/snp), [1000 Genomes](https://www.internationalgenome.org/home), [gnomAD](https://gnomad.broadinstitute.org), [GME](https://illumina.github.io/NirvanaDocumentation/data-sources/gme), [HGSVC](https://www.internationalgenome.org/data-portal/data-collection/hgsvc3) and [DGV](https://dgv.tcag.ca/dgv/app/home)

**Description:**  
PanScan requires these curated BED databases for identifying **novel variants**, **unique sequence regions**, and for performing **genome annotation** across its analysis modules.  
The dataset includes preprocessed variant catalogs and **ideogram (karyotype) files**, enabling modules such as `novel_seq`, `find_uniq_variants`, and `gene_dup` to accurately annotate, compare, and visualize genomic features.


## Gene-duplication analyses

There are 2 parts to this:
You have to first run 
```
panscan make_dup_mtx --gencode_gff3 gencode.gff3 --fofn assemblies.fofn --ref_fa hg38_ref.fa --threads 64
```
to produce the gene-duplication matrix from all your assmeblies.

**Please Note: The gene duplication modules are only compatible with gencode gff3 files**

A sample assembly fofn has been uploaded with the package, it has to follow that format for panscan to use your assemblies. The format is as follows:

```
/path/to/assembly.fa SAMPLE_ID HAPLOTYPE
/path/to/assembly.fa SAMPLE_ID HAPLOTYPE
/path/to/assembly.fa SAMPLE_ID HAPLOTYPE
```
Then the second command 
```
panscan gene_dup --csv_file gene-dup-matrix.csv --ip1 gene_dup/hprc-matrix.csv --ip2 gene_dup/cpc-matrix.csv 
``` 
takes in your gene duplication matrix, and visualizes the duplications in your data and compares them with the hprc and cpc duplications as well. The plots made are :
 - Duplications per assembly
 - Venn diagram of duplications w.r.t HPRC and CPC
 - Frequency comparison of your duplications w.r.t. HPRC and CPC . (Plots the most distinct ones)

**As the HPRC and CPC duplciation matrices have only been shared in HG38, we do not recommend using the comparison plots if you use another reference**

The ```panscan gene_dup``` command will ask you for paths to HPRC and CPC matrices, they are present in the directory you cloned the repo into. Specifically in ```panscan/gene_dup```

To plot an ideogram showing gene duplications use the flag '''--ideogram''' with the command. It also needs the ideo-ip.csv



## Complex regions analyses

**Complex sites** are detected from the VCF file generated by the [Minigraph-Cactus](https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/doc/pangenome.md) pangenome pipeline.  
Following the definition from the draft [Arab Pangenome Reference](https://www.nature.com/articles/s41467-025-61645-w), a *complex site* is defined as:

- A locus with **â‰¥5 haplotypes (alleles)**
- Containing at least **one â‰¥10 kb structural variant (SV)**

You can modify detection thresholds using the following flags:

| Flag | Description | Default |
|------|--------------|----------|
| `-a` | Minimum allele count for a complex site | 5 |
| `-n` | Minimum number of nodes in a site | 1 |
| `-s` | Minimum SV size (bp) | 10000 |

---

**Complex regions** are 100 kb windows that contain at least one complex site and at least one additional structural variant.  
Use the following parameters to define region detection thresholds:

| Flag | Description | Default |
|------|--------------|----------|
| `-l` | Length of the region in bp | 100000 |
| `--sites` | Minimum number of complex sites | 1 |
| `--sv` | Minimum number of secondary SVs | 1 |
| `--threads` | Number of parallel threads to use | 8 |

Activate complex region detection with the `--regions` flag.

---

### End-to-End Run

Run `panscan complex` to identify complex regions and generate haplotype walks and colored GFA graphs for each sample:

```bash
panscan complex \
  --ref_fasta chm13v2.0.fa \
  --gaf_file chm13_mapped_genes.gaf \
  --gff3 chm13v2.0_RefSeq_Liftoff_v5.1.gff3 \
  --sep_pattern '#0#' \
  --ref_name CHM13 \
  -a 5 -n 1 -s 10000 --regions -l 100000 --sites 1 --sv 1 \
  --threads 8 \
  --plot_complex panscan.vcf panscan.gfab
```

This will produce, for each detected region:

- Colored **GFA graphs** for all and protein-coding genes  
- Per-sample **haplotype walk GFAs**  
- Bandage plots with **legends** showing distinct gene colors  
- A detailed `complex_sites_summary.csv` summarizing:
  - Genes detected  
  - Total protein-coding genes  
  - Sample-wise haplotype counts  
  - Region-level statistics

> **Tip:** Add the flag `--no_images` to skip image generation and only produce all summary statistics and data tables.  
> This saves time when manually inspecting regions without rendering Bandage plots.


## Pangenome VCF Processing

For all modules in this section you can use the Sample.vcf file provided in the repo for testing

###  Novel Variant Identification 

This module identifies **novel variants** â€” including **SNPs**, **InDels**, and **SVs** â€” in the input Pangenome VCF by comparing them against public databases such as **dbSNP**, **gnomAD**, **1000 Genomes**, **GME**, and **DGV**.

#### **Usage**
```bash
panscan find_uniq_variants \
  -i /path/to/Sample.vcf \
  -t <VARIANT_TYPE> \
  --db ALL \
  --overlap 80 \
  --db_path /path/to/databases \
  --outdir /path/to/output
```

#### **Variant Types**
- `-t SNP`â€ƒâ†’â€ƒIdentify novel **single-nucleotide polymorphisms**
- `-t INDEL`â€ƒâ†’â€ƒIdentify novel **insertions and deletions**
- `-t SV`â€ƒâ†’â€ƒIdentify novel **structural variants**

#### **Notes**
- Each run automatically creates a unique result folder (e.g. `SV_Comparison_Result`, `SV_Comparison_Result_1`, â€¦) in the specified output directory to avoid overwriting previous results.  
- The `--db_path` argument **must point to the decompressed database directory**, available in the [`Panscan Zenodo Database`](https://zenodo.org/records/17358559) (`database.tar.gz`).  
- When executed inside a Singularity container, all output and temporary files are written to the **calling directory** or the path specified via `--outdir` â€” never inside the container image.


### Novel Seq

This module identifies novel sequences present in a pangenome VCF file (`file1`) by comparing SV insertions against those in another pangenome VCF file (`file2`) and reports them in FASTA format.

#### Overview
1. **Pre-processing**  
   When provided with raw VCFs, the module first normalizes them by splitting multi-allelic variants and decomposing complex variants.  
   If preprocessed VCFs are supplied, this step is skipped.

2. **Comparison**  
   Novel SV insertions present only in `file1` are identified by comparing them against `file2`.

3. **Clustering and Output**  
   Novel insertions detected at similar loci are clustered, and representative novel sequences are written to a FASTA file.

4. **Visualization**  
   Ideogram-style figures are generated to show the genomic distribution of novel insertions, based on chromosome information from the provided database path.

---

#### Usage

```bash
panscan novel_seq \
  -i /path/to/query.vcf \
  -r /path/to/reference.vcf \
  --exclude GRCh38 \
  --db_path /path/to/panscan_db \
  -t 16 \
  --dt 1 \
  --dpi 600
```

Using preprocessed VCFs:

```bash
panscan novel_seq \
  -pInp /path/to/preprocessed_query.vcf \
  -pRef /path/to/preprocessed_reference.vcf \
  --exclude GRCh38 \
  --db_path /path/to/panscan_db \
  -t 16 \
  --dt 1 \
  --dpi 600
```

#### Notes
- The `--db_path` argument **must** point to a valid directory containing `karyotype.bed` [Present in databases.tar in [Zenodo](https://zenodo.org/records/17358559)].  
- If the output folder `NovelSeq_Results` already exists, the program automatically increments the folder name.  
- Preprocessing large pangenome VCFs (like HPRC or CPC) can be time-consuming. Use preprocessed VCFs from [Zenodo](https://zenodo.org/records/17358559) when available.  
- The default number of threads for the decomposition step is set to `1` (recommended). Increasing this value requires sufficient system memory.


## ðŸ“„ Citation
If you use **PanScan** in your work, please cite the paper:

ðŸ‘‰ [Balan, B., Hanif, S., Hashmi, M.A., Kumail, M., BinEshaq, S., Al-Yazeedi, T., Tambi, R., Murtaza, M., Jamalalail, B., Elsokary, H. and Obathani, M.A., 2025. PanScan: A tertiary analysis tool for pangenome graph. bioRxiv, pp.2025-05.
](https://doi.org/10.1101/2025.05.01.651685)




